### PR TITLE
First pass of MAIRL

### PR DESCRIPTION
**Pull request by lalitlal**
**Major In progress**
1. Create Reward Fn and Value Fn in Discriminator [DONE]
- Inside init() or forward() -> added inside forward()

2. [discriminator_irl.py] - modified discriminator.py [DONE]
- Calculate log_p_tau, log_q_tau, log_pq for external losses
- Majority of this is a copy from discriminator.py, but pulled into its own class

3. [mgail.py] - modify how our discriminator is changed [DONE]
- See #2. Discriminator section

4. [mgail.py] - modify al_loss to take into account new [DONE] discriminator output

**TO DO**
- [x] We need to make sure changes to al_loss are appropriate (specifically, the labels)
- [x] Testing that we didn't break existing code 
- [x] Testing the MAIRL actually compiles
- [x] Testing that MAIRL actually works as expected
- [x] Modify ER to account for new field 'action_probs' 
- [x] Modify expert data to include expert action probs  (need for lprobs)
- [x] Copy disabled ant env from AIRL into MGAIL for testing (related to item 4 in this list)

### CODE REVIEW

File: mgail.py [hunk 1]
```diff
@@ -23,19 +24,35 @@ def __init__(self, environment):
         self.temp = tf.compat.v1.placeholder("float", shape=(), name='temperature')
         self.noise = tf.compat.v1.placeholder("float", shape=(), name='noise_flag')
         self.do_keep_prob = tf.compat.v1.placeholder("float", shape=(), name='do_keep_prob')
+        self.lprobs = tf.compat.v1.placeholder('float', shape=(None, 1), name='log_probs')
+        self.nstates = tf.compat.v1.placeholder("float", shape=(None, self.env.state_size), name='states')
```
Comment (yipsang) [inline mgail.py]: I think in the original mgail setup, `states` and `states_` are used interchangeably as next states and current states. For example, forward model uses `states_` as current states and `states` as next states, while discriminator uses only `states` as current states. So we may also reuse the existing placeholder. Or I am also down to have a new placeholder, coz then it's clearer, but be careful that `states` and `states_` are both normalized at https://github.com/yipsang/mgail/blob/master/mgail.py#L63-L64 so we also need to normalize `nstates`.

Comment (yipsang) [reply mgail.py]: But I just saw that we are actually feeding next states to `states` instead of `nstates`, so sounds like this is not used now?

File: mgail.py [hunk 2]
```diff
@@ -75,28 +92,52 @@ def __init__(self, environment):
...
-        discriminator_loss = tf.reduce_mean(d_loss_weighted)
-        self.discriminator.train(objective=discriminator_loss)
+        lprobs = self.lprobs
+        
+        # MODIFIED DISCRIMINATOR SECTION
+        if self.use_irl:
+            self.discrim_output, log_p_tau, log_q_tau, log_pq = self.discriminator.forward(states_, actions, states, lprobs)
+
+
+            correct_predictions = tf.equal(self.discrim_output, tf.argmax(labels, 1))
```
Comment (yipsang) [inline mgail.py]:
```suggestion
            correct_predictions = tf.equal(tf.round(self.discrim_output), tf.argmax(labels, 1))
```
Should we round the discriminator output before comparing them? coz the output should be a scalar ~ [0, 1]? Not sure how to measure it for continuous action though

File: mgail.py [hunk 3]
```diff
@@ -75,28 +92,52 @@ def __init__(self, environment):
...
+
+
+        else:
+            d = self.discriminator.forward(states, actions)
+            # 2.1 0-1 accuracy
+            correct_predictions = tf.equal(tf.argmax(d, 1), tf.argmax(labels, 1))
+            self.discriminator.acc = tf.reduce_mean(tf.cast(correct_predictions, "float"))
+            # 2.2 prediction
+            # d_cross_entropy = tf.nn.softmax_cross_entropy_with_logits(logits=d, labels=labels)
+            d_cross_entropy = tf.nn.softmax_cross_entropy_with_logits(logits=d, labels=labels)
```
Comment (yipsang) [inline mgail.py]:
```suggestion
            d_cross_entropy = tf.nn.softmax_cross_entropy_with_logits(logits=d, labels=labels)
```

File: mgail.py [hunk 4]
```diff
@@ -109,8 +150,16 @@ def policy_loop(state_, t, total_cost, total_trans_err, _):
                 action = common.gumbel_softmax_sample(logits=mu, temperature=self.temp)
 
             # minimize the gap between agent logit (d[:,0]) and expert logit (d[:,1])
-            d = self.discriminator.forward(state_, action, reuse=True)
-            cost = self.al_loss(d)
+
+            # MODIFIED DISCRIMINATOR SECTION:
+            if self.use_irl:
+                self.discrim_output, log_p_tau, log_q_tau, log_pq = self.discriminator.forward(state_, action, self.lprobs, reuse=True)
```
Comment (yipsang) [inline mgail.py]: we need to pass lprobs to the feed_dict when training the policy

File: mgail.py [hunk 5]
```diff
@@ -144,12 +193,15 @@ def policy_stop_condition(state_, t, cost, trans_err, env_term_sig):
...
-        logit_agent, logit_expert = tf.split(axis=1, num_or_size_splits=2, value=d)
+    def al_loss(self, d=None, log_p=None, log_q=None, log_pq=None):
+        if not self.use_irl:
+            logit_agent, logit_expert = tf.split(axis=1, num_or_size_splits=2, value=d)
+            labels = tf.concat([tf.zeros_like(logit_agent), tf.ones_like(logit_expert)], 1)
+            d_cross_entropy = tf.nn.softmax_cross_entropy_with_logits(logits=d, labels=labels)
+        
+        else: # USING IRL
+            labels = tf.concat([1 - self.label, self.label], 1) # TODO: We need to fix this! 
+            d_cross_entropy = labels*(log_p-log_pq) + (1-labels)*(log_q-log_pq)
```
Comment (yipsang) [inline mgail.py]: I think we don't need the labels here. Also, when training the policy, we need to pass lprobs to the feed_dict as well.

I am not sure whether my solution is correct. Please discuss here. I think since we are trying to maximize the reward, so the loss should be the negative reward function. So according to the reward function in page 12,  it should look like this.
```suggestion
            d_cross_entropy = - (log_p - log_pq) + (log_q - log_pq)
```

File: mgail.py (hunk 1 above)
Comment (lalitlal) [reply mgail.py]: So we are using nstates here: https://github.com/yipsang/mgail/pull/7/files#diff-dba3766d7cec900b8de500f888c64a392cd9780f9baf00aae7e3f87a7d3fefc4R92

I think the original discriminator used next states only, so maybe we should use the following (to also take into account normalizing):

`feed_dict = {alg.states_: states_, alg.actions: actions, alg.states: states,
                        alg.label: labels, alg.do_keep_prob: self.env.do_keep_prob,
                        alg.lprobs: lprobs}`

File: mgail.py (hunk 5 above)
Comment (yipsang) [reply mgail.py]: I am thinking the lprobs may not involve expert action_probs but just agent's in this case.

File: mgail.py (hunk 1 above)
Comment (lalitlal) [reply mgail.py]: UPDATE: We are using the states instead of nstates placeholder haha, so I'll just remove nstates placeholder.

File: mgail.py (hunk 2 above)
Comment (lalitlal) [reply mgail.py]: Good catch! Since our discriminator in the IRL case is using a single dim output (1) instead of 2, we'll need to do a round.

File: mgail.py (hunk 4 above)
Comment (lalitlal) [reply mgail.py]: Wait why does the policy need lprobs? Where would it use it?

File: mgail.py (hunk 1 above)
Comment (lalitlal) [reply mgail.py]: Addressed in https://github.com/yipsang/mgail/pull/7/commits/e9397141abebae3119b02e5f33ebfcac924c154e

File: mgail.py (hunk 3 above)
Comment (lalitlal) [reply mgail.py]: Addressed in https://github.com/yipsang/mgail/pull/7/commits/e9397141abebae3119b02e5f33ebfcac924c154e

File: mgail.py (hunk 2 above)
Comment (lalitlal) [reply mgail.py]: Addressed in https://github.com/yipsang/mgail/pull/7/commits/e9397141abebae3119b02e5f33ebfcac924c154e

File: mgail.py (hunk 5 above)
Comment (s-saleem) [reply mgail.py]: I think you may be right about not needing the labels anymore, but not sure what you mean by lprobs may not involve expert action_probs?

I'm thinking the discriminator learns what good action_probs look like from the expert and then scores the generated policy action_probs according to that?

Comment (lalitlal) [reply mgail.py]: I addressed changing the cross entropy loss here:

https://github.com/yipsang/mgail/pull/7/commits/c7402b9ba1a8471aa5dc7443a8046afd0bd29071

But not too sure about lprobs stuff :P let's continue discussion here

File: mgail.py (hunk 4 above)
Comment (s-saleem) [reply mgail.py]: Yea i think doesn't need lprobs as input just the discriminator?

Comment (yipsang) [reply mgail.py]: But lprobs is actually the log_q_tau, which will be needed to calculate the reward. The problem is just whether it should  involve both expert's and agent's or just agent's prob.

Comment (yipsang) [reply mgail.py]: Just in case you don't realize, this line is inside the policy loop, so the lprobs is used in training policy

Comment (yipsang) [reply mgail.py]: or maybe the lprobs in this line should be the log(action)?

Comment (lalitlal) [reply mgail.py]: In AIRL, they use expert and agent lprobs:

`lprobs_batch = np.expand_dims(np.concatenate([lprobs_batch, expert_lprobs_batch], axis=0), axis=1).astype(np.float32)`

File: driver.py [hunk 6]
```diff
@@ -149,11 +149,13 @@ def collect_experience(self, record=1, vis=0, n_steps=None, noise_flag=True, sta
             if not noise_flag:
                 do_keep_prob = 1.
             
-            a, a_probs = self.sess.run(fetches=[alg.action_test, alg.action_probs], feed_dict={alg.states: np.reshape(observation, [1, -1]),
+            a, a_means = self.sess.run(fetches=[alg.action_test, alg.action_means], feed_dict={alg.states: np.reshape(observation, [1, -1]),
                                                                     alg.do_keep_prob: do_keep_prob,
                                                                     alg.noise: noise_flag,
                                                                     alg.temp: self.env.temp})
 
+            a_probs = common.compute_action_probs(a, a_means[0], np.repeat(self.env.sigma, a.shape[0]))
```
Comment (lalitlal) [inline driver.py]: don't we need to use all the a_means instead of a_means[0]? 

The action taken is a vector [a,b,c], each with their own mean and log_std right?

File: mgail.py [hunk 7]
```diff
@@ -186,6 +171,21 @@ def policy_loop(state_, t, total_cost, total_trans_err, _):
             total_trans_err += tf.reduce_mean(abs(nu))
             t += 1
 
+            # minimize the gap between agent logit (d[:,0]) and expert logit (d[:,1])
```
Comment (lalitlal) [inline mgail.py:174]: Just curious, what was the reason for moving this down?

File: driver.py (hunk 6 above)
Comment (s-saleem) [reply driver.py]: I think this the issue here was that the a_means was returned as 2d array, (1, actions_len)
